colcon_override_check/package_selection/override_check.py

```diff
@@ -0,0 +1,171 @@
+
+        parser.add_argument(
+            '--allow-overriding',
+            action='extend',
+            default=[],
+            metavar='PKG_NAME',
+            nargs='+',
+            help='Allow building packages that exist in underlay workspaces')
+
+    def check_parameters(self, args, pkg_names):  # noqa: D102
+        for pkg_name in args.allow_overriding or []:
+            if pkg_name not in pkg_names:
+                logger.warning(
+                    "ignoring unknown package '{pkg_name}' in "
+                    '--allow-overriding'.format_map(locals()))
+
+    def select_packages(self, args, decorators):  # noqa: D102
+        verb_name = getattr(args, 'verb_name', None)
+        if verb_name not in OverrideCheckPackageSelection.TARGET_VERBS:
+            return
+
+        # Enumerate the names of packages selected to be built.
+        selected_names = {d.descriptor.name for d in decorators if d.selected}
+
+        # Enumerate the underlay packages.
+        identification_extensions = get_package_identification_extensions()
+        underlay_descriptors = discover_packages(
+            args, identification_extensions)
+        augment_packages(underlay_descriptors)
+        if not underlay_descriptors:
+            return
+
+        # Enumerate workspace packages already built (if applicable).
+        building_or_built = set(selected_names)
+        if hasattr(args, 'install_base'):
+            install_base = Path(args.install_base).resolve()
+
+            # Drop any enumerated underlay packages in our install base.
+            # This could happen when a workspace install base is activated and
+            # then built again.
+            rebuilding = [
+                d for d in underlay_descriptors if d.path == install_base]
+            underlay_descriptors.difference_update(rebuilding)
+            if not underlay_descriptors:
+                return
+            building_or_built.update(d.name for d in rebuilding)
+
+            # Explicitly enumerate which workspace packages have already been
+            # built/installed in case the install base hasn't already been
+            # activated.
+            index_path = install_base / get_relative_package_index_path()
+            if index_path.is_dir():
+                building_or_built.update(p.name for p in index_path.iterdir())
+
+        # Enumerate dependencies from underlay packages to selected packages.
+        reverse_dependencies = {}
+        for d in underlay_descriptors:
+            # If an underlay package is being built or has already been built,
+            # its dependencies are irrelevant because it is already, or will
+            # be, overridden itself.
+            if d.name in building_or_built:
+                continue
+
+            # Consider only runtime dependencies from underlay packages to
+            # selected packages.
+            deps = {
+                d.name for d in d.get_dependencies(categories=('run',))
```

Comment:
Are build_export dependencies put into the `run` category? I think this case below only depends on an underlay descriptor having a build and build_export dependency on the overridden package.

Say one of the selected names, which is being overridden, is a header only library (Let's call it **foo**). Say it contains a struct definition, and that definition reordered or changed it's members between the underlay and overlay versions of **foo**. Say there's a package in the underlay that has a build and build_export dependency to **foo** (Let's call it **bar**), and it has an api `bar_with_me(foo::SomeStruct s)` which accepts an instance of a struct from **foo**. Say another package in the overlay depends on both **foo** and **bar** (Let's call it **baz**). Say **baz** uses **foo**'s API to instantiate a struct, and passes that to `bar_with_me(..)`. I think that's undefined behavior.

Reply:
At the moment, the `<build_export_depends>` entries end up as `run` dependencies in colcon ([here's the spot](https://github.com/colcon/colcon-ros/blob/07c7e0622cd9eb4749a2ae1031da95fb405cb20c/colcon_ros/package_identification/ros.py#L79-L87)).

So you should get a warning in your example, if I understand correctly.